Sources/Regex/Parse/Parse.swift

```diff
@@ -183,19 +215,56 @@ extension Parser {
       return .symmetricDifference
     }
   }
+*/
 
-  ///     CharacterClass -> '[' CharacterSetComponent+ ']'
+  /// Parse a custom character class
   ///
-  ///     CharacterSetComponent -> CharacterSetComponent SetOp CharacterSetComponent
-  ///     CharacterSetComponent -> CharacterClass
-  ///     CharacterSetComponent -> <token: Character>
-  ///     CharacterSetComponent -> <token: Character> '-' <token: Character>
+  ///     CustomCharClass -> Start Set (SetOp Set)* ']'
+  ///     Set             -> Member+
+  ///     Member          -> CustomCharClass | !']' !SetOp (Range | Atom)
+  ///     Range           -> Atom `-` Atom
   ///
-  mutating func parseCustomCharacterClass() throws -> CharacterClass {
-    try lexer.eat(expecting: .leftSquareBracket)
-    let isInverted = lexer.tryEat(.caret)
-    var components: [CharacterSetComponent] = []
-    while !lexer.tryEat(.rightSquareBracket) {
+  mutating func parseCustomCharacterClass(
+    _ start: CustomCharacterClass.Start
+  ) throws -> CustomCharacterClass {
+    typealias Member = CustomCharacterClass.Member
+    var members: Array<Member> = []
+
+    // TODO: Is this a correct/sane associativity? Precedence?
```

Comment:
Left association by default makes sense to me (especially for subtraction). We may want to give some operators precedence over others, though it may be worth getting users to spell it out explicitly

Reply:
Alrighty, I will consider this concern deferred to you :-)